### PR TITLE
Parser: fix ambiguity with whitespace in version ranges

### LINKED ISSUE #40342: Spack add parser: Unbounded version specifier followed by another spec
### Steps to reproduce

An error occur when running this command:
```spack add opa-psm2^cuda@10.2: nccl
Error: 10.2:nccm is an empty range
```
You can reproduce this whenever you have 2 specs to add, use unbounded version specifier as last part of the first spec. 
It seems like the parser does not treat nccl in this case as a new spec but as the upper bound of the version.

Also, for some reason the last character of the second spec is incremented like in this example: nccl become nccm.
When I used package opa-psm2 instead of nccl, the error would be: Error: 10.2:opa-psm3 is an empty range

### Error message

_No response_

### Information on your system

* **Spack:** 0.21.0.dev0 (38dc3a6896f64b894adead3446e8f822f878860b)
* **Python:** 3.6.5
* **Platform:** linux-rhel8-cascadelake
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

### PR DESCRIPTION
**Pull request by haampie**
Fixes #40342

This is a breaking change, but imho `develop` is broken, so it's rather a bugfix.

Allowing white space around `:` in version ranges introduces an ambiguity:

```
a@1: b
```

parses as `a@1:b` but should really be parsed as two separate specs `a@1:` and `b`.

With white space disallowed around `:` in ranges, the ambiguity is resolved.

---

White space is still allowed around the `,` separator in version lists, since there we unambiguously require a next version/range token anyways, and in fact the syntax with whitespaces around `,` is used in a bunch of packages.

---

Apart from that, there's this rather odd, **supported** syntax:

```
a@1:b = c
```

It's parsed as `[a, @1:, b=c]`. Apparently this syntax is relied on in *exactly one* package -- I think disallowing that would have made sense if it wasn't relied on... I've left this as is, and added an additional test for this syntax.

### CODE REVIEW

File: lib/spack/spack/test/spec_syntax.py [hunk 1]
```diff
@@ -472,37 +472,17 @@ def _specfile_for(spec_str, filename):
...
-            ],
-            "%intel@12.1:12.6+debug",
-        ),
-        (
-            "@ 12.1 : 12.6 + debug - qt_4",
+            "a@1: b",
             [
-                Token(TokenType.VERSION, value="@ 12.1 : 12.6"),
-                Token(TokenType.BOOL_VARIANT, value="+ debug"),
-                Token(TokenType.BOOL_VARIANT, value="- qt_4"),
```
Comment (alalazo) [inline lib/spack/spack/test/spec_syntax.py]: Can we keep these two tests and just remove the versions? I would be fine not allowing spaces after a sigil, but until we do that we should probably test variants and compilers.

**Review by alalazo — CHANGES_REQUESTED** [review]
LGTM. Only a minor request.

File: lib/spack/spack/test/spec_syntax.py (hunk 1 above)
Comment (haampie) [reply lib/spack/spack/test/spec_syntax.py]: Yeah, agreed

File: lib/spack/spack/parser.py [hunk 2]
```diff
@@ -96,7 +96,7 @@
 QUOTED_VALUE = r"[\"']+([a-zA-Z_0-9\-+\*.,:=\~\/\\\s]+)[\"']+"
 
 VERSION = r"=?([a-zA-Z0-9_][a-zA-Z_0-9\-\.]*\b)"
-VERSION_RANGE = rf"({VERSION}\s*:\s*{VERSION}(?!\s*=)|:\s*{VERSION}(?!\s*=)|{VERSION}\s*:|:)"
+VERSION_RANGE = rf"({VERSION}:{VERSION}(?!=)|:{VERSION}(?!=)|{VERSION}:|:)"
```
Comment (haampie) [inline lib/spack/spack/parser.py]:
```suggestion
VERSION_RANGE = rf"(({VERSION})?:({VERSION})?)"
```

IMHO negative lookahead can now also go.

Under the current rules `@:abc=def` parses as `@: abc=def`, but maybe better to parse a a range `@:abc` followed by "unexpected =" syntax error?

**Comment by tgamblin** [comment]
> Apparently this syntax is relied on in exactly one package 

Which package relies on this?

**Comment by haampie** [comment]
https://github.com/spack/spack/blob/e2a7f2ee9acdacd6bdd78cf12ac7fe1a15c578ca/var/spack/repos/builtin/packages/py-wasabi/package.py#L19-L23

**Review by alalazo — APPROVED** [review]
LGTM, but let's wait for at least another :heavy_check_mark: for the breaking change.

**Review by becker33 — DISMISSED** [review]
Can you add this detail to the comment at the top of `lib/spack/spack/parser.py`? I think that's the only relevant documentation to update.

**Review by tgamblin — CHANGES_REQUESTED** [review]
Since this is breaking anyway I don't think it's unreasonable to *not* support the `py-wasabi` case.  We generally require a space to differentiate key-value pairs in other places... I don't see why we shouldn't also require it after a version range.

so, disallow:

```
a@1:b = c
```

and require this space:

```
a@1: b = c
    ^
```
Or do folks object?

**Comment by haampie** [comment]
These are all valid:

- `@1:^python`
- `@:2, 4%gcc`
- `build_system    =    makefile%gcc`

Maybe `%` and `^` are visually good enough separators, but to me it feels like it's all not very consistent, and disallowing `a@1:b=c` should also mean disallowing all of the above, and that will certainly make it more likely that user repos break

All whitespace in verison specifiers should've been banned from the start ;p

**Comment by tgamblin** [comment]
I don’t think it’s inconsistent. In other places we change context based on sigils, and require space for kvp’s (which start with an identifier). This is just (I think?) the only place where there is a word boundary because of a colon at the *end* of the thing before a kvp.

OTOH, if there is no need to disable support for it (does it introduce ambiguity or just weirdness?) then it could stay like it is.

**Comment by haampie** [comment]
It's not ambiguous, `@1:a=b` cannot be parsed as `[@1:a, =b]`

**Comment by haampie** [comment]
> Can you add this detail to the comment at the top of `lib/spack/spack/parser.py`? I think that's the only relevant documentation to update.

What do you want to add exactly? Those high-level rules are generally unclear about use of white space, as I read them they're even more strict than what we allow, but it's hard to follow since it's not valid EBNF. I think the implementation + tests are the only correct documentation

**Review by tgamblin — APPROVED** [review]

**Comment by tgamblin** [comment]
@haampie: I think my concerns are addressed by the conversation but we're still waiting on @becker33's.

**Comment by haampie** [comment]
Ping @becker33

**Comment by haampie** [comment]
Ping @becker33

**Review by alalazo — APPROVED** [review]
